app/users/wallets.go

```diff
@@ -0,0 +1,111 @@
+package users
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/lbryio/lbrytv/internal/lbrynet"
+	"github.com/lbryio/lbrytv/internal/monitor"
+	"github.com/lbryio/lbrytv/models"
+	"github.com/sirupsen/logrus"
+	"github.com/volatiletech/sqlboiler/boil"
+)
+
+type WalletService struct {
+	UserService
+}
+
+// NewWalletService returns UserService instance for retrieving or creating wallet-based user records and accounts.
+func NewWalletService() *WalletService {
+	s := &WalletService{UserService{logger: monitor.NewModuleLogger("users")}}
+	return s
+}
+
+func (s *WalletService) Retrieve(q Query) (*models.User, error) {
+	var (
+		localUser *models.User
+		wid       string
+	)
+	token := q.Token
+
+	log := s.logger.LogF(monitor.F{"token": token})
+
+	remoteUser, err := getRemoteUser(token, q.MetaRemoteIP)
+	if err != nil {
+		return nil, s.LogErrorAndReturn(log, "cannot authenticate user with internal-apis: %v", err)
+	}
+
+	// Update log entry with extra context data
+	log = s.logger.LogF(monitor.F{"token": token, "id": remoteUser.ID, "email": remoteUser.Email})
+	if remoteUser.Email == "" {
+		return nil, s.LogErrorAndReturn(log, "cannot authenticate user with internal-api, email not confirmed")
+	}
+
+	localUser, errStorage := s.getDBUser(remoteUser.ID)
+	if errStorage == sql.ErrNoRows {
+		log.Infof("user was not found in the database, creating")
+		localUser, err = s.createDBUser(remoteUser.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		wid, err = s.createWallet(localUser)
+		if err != nil {
+			return nil, err
+		}
+		log = s.logger.LogF(monitor.F{"token": token, "id": remoteUser.ID, "email": remoteUser.Email, "wallet_id": wid})
+	} else if errStorage != nil {
+		return nil, errStorage
+	}
+
+	// This scenario may happen for legacy users who haven't moved to wallets yet
+	if localUser.WalletID == "" {
+		log.Warn("user doesn't have wallet ID set")
+		wid, err = s.createWallet(localUser)
+		if err != nil {
+			return nil, err
+		}
+		err := s.saveWalletID(localUser, wid)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return localUser, nil
+}
+
+func (s *WalletService) createWallet(u *models.User) (string, error) {
```

Comment:
To create a consistent interface with other user entities manipulation methods and to avoid sprinkling semi-external dependencies in the main `Retrieve` method.